pkg/debuginfo/store.go

```diff
@@ -192,87 +205,118 @@ func (s *Store) Upload(stream debuginfopb.DebugInfoService_UploadServer) error {
 	level.Debug(s.logger).Log("msg", "trying to upload debug info", "buildid", buildID)
 
 	ctx := stream.Context()
-	if result, err := s.metadataManager.fetch(ctx, buildID); err == nil {
-		level.Debug(s.logger).Log("msg", "fetching metadata state", "result", result)
+	metadataFile, err := s.metadataManager.fetch(ctx, buildID)
+	if err == nil {
+		level.Debug(s.logger).Log("msg", "fetching metadata state", "result", metadataFile)
 
-		switch result {
-		case metadataStateEmpty:
-			// Not created yet.
+		switch metadataFile.State {
+		case metadataStateCorrupted:
+			// Corrupted. Re-upload.
 		case metadataStateUploaded:
 			// The debug info was fully uploaded.
-			fallthrough
+			return status.Error(codes.AlreadyExists, "debuginfo already exists")
 		case metadataStateUploading:
-			return status.Error(codes.AlreadyExists, "debuginfo already exists, being uploaded right now")
+			if !isStale(metadataFile) {
+				return status.Error(codes.AlreadyExists, "debuginfo already exists, being uploaded right now")
+			}
+			// The debug info upload operation most likely failed.
+		default:
+			return status.Error(codes.Internal, "unknown metadata state")
 		}
 	} else {
-		level.Error(s.logger).Log("msg", "failed to fetch metadata state", "err", err)
+		if !errors.Is(err, ErrMetadataNotFound) {
+			level.Error(s.logger).Log("msg", "failed to fetch metadata state", "err", err)
+		}
 	}
 
 	found, err := s.find(ctx, buildID)
 	if err != nil {
 		return err
 	}
 
+	var objFileHash string
 	if found {
+		if req.GetInfo().Hash != "" {
+			if metadataFile != nil && metadataFile.Hash != "" {
```

Comment:
nit: we don't need the `metadataFile.Hash != ""` check as we will check if it's equal to the sent hash below, and it can't be an empty string